projectq/libs/math/_gates.py

```diff
@@ -200,19 +204,361 @@ def __init__(self, a, N):
         It also initializes its base class, BasicMathGate, with the
         corresponding function, so it can be emulated efficiently.
         """
-        BasicMathGate.__init__(self, lambda x: ((a * x) % N,))
+        BasicMathGate.__init__(self, lambda x: ((a * x) % N, ))
         self.a = a
         self.N = N
 
     def __str__(self):
         return "MultiplyByConstantModN({}, {})".format(self.a, self.N)
 
     def __eq__(self, other):
-        return (isinstance(other, MultiplyByConstantModN) and
-                self.a == other.a and self.N == other.N)
+        return (isinstance(other, MultiplyByConstantModN) and self.a == other.a
+                and self.N == other.N)
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class AddQuantumGate(BasicMathGate):
+    """
+    Adds up two quantum numbers represented by quantum registers.
+    The numbers are stored from low- to high-bit, i.e., qunum[0] is the LSB.
+
+    Example:
+        .. code-block:: python
+
+            qunum_a = eng.allocate_qureg(5) # 5-qubit number
+            qunum_b = eng.allocate_qureg(5) # 5-qubit number
+            carry_bit = eng.allocate_qubit()
+
+            X | qunum_a[2] #qunum_a is now equal to 4
+            X | qunum_b[3] #qunum_b is now equal to 8
+            AddQuantum | (qunum_a, qunum_b, carry)
+            # qunum_a remains 4, qunum_b is now 12 and carry_bit is 0
+    """
+    def __init__(self):
+        BasicMathGate.__init__(self, None)
+
+    def __str__(self):
+        return "AddQuantum"
+
+    def __eq__(self, other):
+        return (isinstance(other, AddQuantumGate))
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def get_math_function(self, qubits):
+        n = len(qubits[0])
+
+        def math_fun(a):
+            a[1] = a[0] + a[1]
+            if len(bin(a[1])[2:]) > n:
+                a[1] = a[1] % (2**n)
+
+                # The part below was originally written by peter-janderks
```

Comment:
This part was not entirely clear to me...

Reply:
This does make things clearer to me now. I think I'll rewrite this once again to make this intention clearer in the code.

Reply:
This also means, that in the documentation string above, the carry qubit needs to be a quantum register with 2 qubits if we want to be consistent.